### PR TITLE
Syntax Error: add message with the plugin that should be enabled

### LINKED ISSUE #169: Better error message on `unexpected token` when it's because a plugin isn't turned on
Ref https://github.com/babel/babylon/pull/163#discussion_r82658832

Example `if (!this.hasPlugin("dynamicImport")) this.unexpected();` or

```
Module build failed: SyntaxError: Unexpected token (1:6)

> 1 | import();
    |       ^
  2 |
```

If a user is using an experimental syntax and we error then we should tell them a message so it can be fixed.

We don't necessarily know they are using babel/babylon though so the error message can give some hints:
- ask the user to turn on the corresponding plugin in babylon
- ask the user to add the syntax/transform plugin in babel
- ask the user to use `parserOpts: { plugins: ['syntax'] }`

### PR DESCRIPTION
**Pull request by hzoo**
Fixes #169
Closes #178
Rebased/rewrite of https://github.com/babel/babylon/pull/178

Add a new method: `this.expectPlugin` instead of `this.hasPlugin` which throws if the plugin isn't enabled.

screen shot 2017-08-03 at 5 00 49 pm

This also means something better than "unexpected token" when it's a config issue. This also isn't tied to Babel but just babylon's plugins, so other tools can catch the error like Babel will to produce their own specific error messages.

screen shot 2017-08-03 at 5 01 51 pm

- [ ] non-block: make sure this covers all the relevant uses of `hasPlugin` (can be in later PRs)
  - [ ] bigInt, classPrivateProperties, exportExtensions, numericSeparator
- [ ] document the need to write a test without a plugin to get test coverage for each plugin added in the future

### CODE REVIEW

File: src/parser/statement.js [hunk 1]
```diff
@@ -491,12 +486,13 @@ export default class StatementParser extends ExpressionParser {
     if (this.match(tt._catch)) {
       const clause = this.startNode();
       this.next();
-      if (this.match(tt.parenL) || !this.hasPlugin("optionalCatchBinding")) {
+      if (this.match(tt.parenL)) {
         this.expect(tt.parenL);
```
Comment (Kovensky) [inline src/parser/statement.js]: I guess this became redundant now?

File: src/parser/expression.js [hunk 2]
```diff
@@ -627,11 +622,11 @@ export default class ExpressionParser extends LValParser {
         return this.finishNode(node, "Super");
 
       case tt._import:
-        if (this.hasPlugin("importMeta") && this.lookahead().type === tt.dot) {
+        if (this.lookahead().type === tt.dot) {
           return this.parseImportMetaProperty();
```
Comment (bakkot) [inline src/parser/expression.js]: Did you drop a `this.expectPlugin("importMeta")` here?

Comment (nicolo-ribaudo) [inline src/parser/expression.js]: It's called by the `parseImportMetaProperty` function.

Comment (hzoo) [inline src/parser/expression.js]: Yeah this problem was brought up by @motiz88, @danez in https://github.com/babel/babylon/pull/178/files#r87556780

If we make the error here I guess it's a different error if they just have `import.` vs `import.asdf` but maybe we shouldn't make a big deal out of it?

File: src/parser/expression.js [hunk 3]
```diff
@@ -679,7 +674,9 @@ export default class ExpressionParser extends LValParser {
         return id;
 
       case tt._do:
-        if (this.hasPlugin("doExpressions")) {
+        // TODO
+        if (true) {
```
Comment (danez) [inline src/parser/expression.js]: What is this ? :D

File: src/parser/expression.js [hunk 4]
```diff
@@ -808,6 +801,16 @@ export default class ExpressionParser extends LValParser {
     propertyName: string,
   ): N.MetaProperty {
     node.meta = meta;
+
+    if (meta.name === "function" && propertyName === "sent") {
+      if (this.isContextual(propertyName)) {
+        this.expectPlugin("functionSent");
+      } else if (!this.hasPlugin("functionSent")) {
+        // They didn't actually say `function.sent`, just `function.`, so a simple error would be less confusing.
+        this.unexpected();
```
Comment (danez) [inline src/parser/expression.js]: Uncovered. I add a test.

File: src/parser/expression.js (hunk 3 above)
Comment (hzoo) [inline src/parser/expression.js]: there's a bug somewhere if I don't have it if you want to look into that 😄

Comment (Kovensky) [inline src/parser/expression.js]: Is this `if` necessary at all? The `_do` context can only be legally entered if the plugin is enabled to begin with.

**Comment by danez** [comment]
I enabled the no-case-declarations. This should prevent this problems.

**Comment by danez** [comment]
Okay I removed the if (true) and also rebased the branch.

File: src/parser/expression.js (hunk 3 above)
Comment (hzoo) [inline src/parser/expression.js]: no it's not yeah I just need a block if that works

File: src/parser/statement.js [hunk 5]
```diff
@@ -1145,17 +1145,14 @@ export default class StatementParser extends ExpressionParser {
...
+    if (!node.typeAnnotation) {
+      this.expectOnePlugin(["classProperties"]);
     }
 
     this.state.inClassProperty = true;
 
     if (this.match(tt.eq)) {
-      if (!this.hasPlugin("classProperties"))
-        this.raise(this.state.start, noPluginMsg);
+      this.expectOnePlugin(["classProperties"]);
```
Comment (nicolo-ribaudo) [inline src/parser/statement.js]: Here you can use `exceptPlugin`

File: src/parser/statement.js [hunk 6]
```diff
@@ -1145,17 +1145,14 @@ export default class StatementParser extends ExpressionParser {
   }
 
   parseClassProperty(node: N.ClassProperty): N.ClassProperty {
-    const noPluginMsg =
-      "You can only use Class Properties when the 'classProperties' plugin is enabled.";
-    if (!node.typeAnnotation && !this.hasPlugin("classProperties")) {
-      this.raise(node.start, noPluginMsg);
+    if (!node.typeAnnotation) {
+      this.expectOnePlugin(["classProperties"]);
```
Comment (nicolo-ribaudo) [inline src/parser/statement.js]: Here you can use `exceptPlugin`

File: test/fixtures/experimental/_no-plugin/class-properties-with-initializer/options.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
+  "throws": "This experimental syntax requires enabling one of the following parser plugin(s): 'classProperties, typescript' (2:6)",
```
Comment (nicolo-ribaudo) [inline test/fixtures/experimental/_no-plugin/class-properties-with-initializer/options.json]: Is this influenced by #666?

**Pull request merged**